File: frontend/src/services/whiteboard.ts
/**
 * 课堂白板 REST API
 */
import api from './api'

export interface WhiteboardStateDto {
  session_id: number
  cell_id: number
  document: Record<string, unknown>
  version: number
}

export interface SessionGroupsDto {
  session_id: number
  groups: Array<{ id: number; group_index: number; label: string }>
  members: Array<{ user_id: number; group_index: number; display_name?: string }>
}

export const whiteboardService = {
  async getState(sessionId: number, cellId: number): Promise<WhiteboardStateDto> {
    const res = await api.get(
      `/classroom-sessions/sessions/${sessionId}/whiteboard/${cellId}/state`
    )
    return res.data
  },

  async setMode(
    sessionId: number,
    cellId: number,
    mode: 'setup' | 'collaborate' | 'locked'
  ): Promise<WhiteboardStateDto> {
    const res = await api.patch(
      `/classroom-sessions/sessions/${sessionId}/whiteboard/${cellId}/mode`,
      { mode }
    )
    return res.data
  },

  async getGroups(sessionId: number): Promise<SessionGroupsDto> {
    const res = await api.get(`/classroom-sessions/sessions/${sessionId}/groups`)
    return res.data
  },

  async setupGroups(
    sessionId: number,
    groupCount: number,
    randomAssign = true
  ): Promise<SessionGroupsDto> {
    const res = await api.put(`/classroom-sessions/sessions/${sessionId}/groups/setup`, {
      group_count: groupCount,
      random_assign: randomAssign,
    })
    return res.data
  },
}

File: backend/app/schemas/whiteboard.py
"""白板与课堂分组 Schemas"""

from datetime import datetime
from typing import Any, Dict, List, Optional

from typing import Literal

from pydantic import BaseModel, Field


class SessionGroupMemberOut(BaseModel):
    user_id: int
    group_index: int
    display_name: Optional[str] = None


class SessionGroupsResponse(BaseModel):
    session_id: int
    groups: List[Dict[str, Any]]
    members: List[SessionGroupMemberOut]


class SetupGroupsRequest(BaseModel):
    group_count: int = Field(..., ge=1, le=12)
    random_assign: bool = True


class PatchMemberGroupRequest(BaseModel):
    group_index: int = Field(..., ge=1, le=12)


class WhiteboardStateResponse(BaseModel):
    session_id: int
    cell_id: int
    document: Dict[str, Any]
    version: int
    updated_at: Optional[datetime] = None


class WhiteboardModeRequest(BaseModel):
    mode: Literal["setup", "collaborate", "locked"]


class WhiteboardOpMessage(BaseModel):
    cell_id: int
    op: Dict[str, Any]
